File: Fbs.WebApi/Repository/BookingRepository.cs
using System.Linq.Expressions;
using Fbs.WebApi.Entities;
using Fbs.WebApi.Options;
using Google.Apis.Calendar.v3;
using Google.Apis.Calendar.v3.Data;
using Microsoft.Extensions.Caching.Hybrid;
using Microsoft.Extensions.Options;

namespace Fbs.WebApi.Repository;

public class BookingRepository(
    InstrumentationSource instrumentation,
    HybridCache cache,
    IFreeSql freeSql,
    IOptions<GoogleOptions> options,
    CalendarService calendarService,
    UserRepository userRepository,
    ILogger<BookingRepository> logger
) : IRepository<Booking>
{
    public async Task<List<Booking>> GetListAsync(CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        return await cache.GetOrCreateAsync("Bookings", freeSql, async (state, ct) =>
        {
            return await state.Select<Booking>().ToListAsync(ct);
        }, cancellationToken: cancellationToken);
    }

    public async Task<Booking?> FindAsync(Expression<Func<Booking, bool>> predicate,
        CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        var list = await GetListAsync(cancellationToken);
        return list.SingleOrDefault(predicate.Compile());
    }

    public async Task<Booking> GetAsync(Expression<Func<Booking, bool>> predicate,
        CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        var list = await GetListAsync(cancellationToken);
        return list.Single(predicate.Compile());
    }

    public async Task<Booking> InsertAsync(Booking entity, CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        var user = await userRepository.GetAsync(u => u.Phone == entity.UserPhone, cancellationToken);

        entity.Id = Guid.NewGuid();

        await freeSql.Insert(entity).ExecuteAffrowsAsync(cancellationToken);

        try
        {
            var @event = CreateCalendarEvent(entity, user);
            await Task.WhenAll([
                calendarService.Events.Insert(@event, options.Value.CarbonCopyCalendarId)
                    .ExecuteAsync(cancellationToken)
            ]);
        }
        catch (Exception ex)
        {
            logger.LogWarning(ex, "Failed to sync booking {BookingId} to Google Calendar carbon copy", entity.Id);
        }

        await cache.RemoveAsync("Bookings", cancellationToken);

        return entity;
    }

    public async Task<Booking> UpdateAsync(Booking entity, CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        var booking = await freeSql.Select<Booking>()
            .Where(b => b.Id == entity.Id)
            .FirstAsync(cancellationToken);

        booking.StartDateTime = entity.StartDateTime;
        booking.EndDateTime = entity.EndDateTime;
        booking.Conduct = entity.Conduct;
        booking.Description = entity.Description;
        booking.FacilityName = entity.FacilityName;
        booking.PocName = entity.PocName;
        booking.PocPhone = entity.PocPhone;
        booking.UserPhone = entity.UserPhone;

        await freeSql.Update<Booking>()
            .SetSource(booking)
            .ExecuteAffrowsAsync(cancellationToken);

        try
        {
            var user = await userRepository.GetAsync(u => u.Phone == booking.UserPhone, cancellationToken);
            var @event = CreateCalendarEvent(booking, user);
            await Task.WhenAll([
                calendarService.Events.Update(@event, options.Value.CarbonCopyCalendarId, booking.Id.ToString("N"))
                    .ExecuteAsync(cancellationToken)
            ]);
        }
        catch (Exception ex)
        {
            logger.LogWarning(ex, "Failed to sync booking {BookingId} update to Google Calendar carbon copy", booking.Id);
        }

        await cache.RemoveAsync("Bookings", cancellationToken);

        return booking;
    }

    public async Task DeleteAsync(Expression<Func<Booking, bool>> predicate,
        CancellationToken cancellationToken = default)
    {
        using var activity = instrumentation.ActivitySource.StartActivity();

        var booking = await GetAsync(predicate, cancellationToken);

        await freeSql.Delete<Booking>()
            .Where(b => b.Id == booking.Id)
            .ExecuteAffrowsAsync(cancellationToken);

        try
        {
            await Task.WhenAll([
                calendarService.Events.Delete(options.Value.CarbonCopyCalendarId, booking.Id.ToString("N"))
                    .ExecuteAsync(cancellationToken)
            ]);
        }
        catch (Exception ex)
        {
            logger.LogWarning(ex, "Failed to sync booking {BookingId} deletion to Google Calendar carbon copy", booking.Id);
        }

        await cache.RemoveAsync("Bookings", cancellationToken);
    }

    private static Event CreateCalendarEvent(Booking booking, Entities.User user)
    {
        return new Event
        {
            Id = booking.Id.ToString("N"),
            Summary = $"{user.Unit} {booking.Conduct}",
            Start = new EventDateTime
            {
                DateTimeDateTimeOffset = booking.StartDateTime,
            },
            End = new EventDateTime
            {
                DateTimeDateTimeOffset = booking.EndDateTime,
            },
            Location = booking.FacilityName,
            Description = $"""
                           Point of contact: {booking.PocName} / {booking.PocPhone}

                           Booked by: {user.Unit} / {user.Name}
                           Number: {user.Phone}

                           Description: 
                           {booking.Description}
                           """,
        };
    }
}

File: Fbs.WebApi/src/main/kotlin/sg/from/fbs/repository/BookingRepository.kt
package sg.from.fbs.repository

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.google.api.client.util.DateTime
import com.google.api.services.calendar.Calendar
import com.google.api.services.calendar.model.Event
import com.google.api.services.calendar.model.EventDateTime
import org.springframework.cache.annotation.CacheEvict
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Repository
import sg.from.fbs.config.GoogleProperties
import sg.from.fbs.entities.Booking
import java.time.OffsetDateTime
import java.util.Base64
import java.util.UUID
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Lazy

@Repository
class BookingRepository(
    private val calendarService: Calendar,
    private val options: GoogleProperties,
    private val userRepository: UserRepository
) {
    @Autowired
    @Lazy
    lateinit var self: BookingRepository

    private val objectMapper: ObjectMapper = jacksonObjectMapper().registerModule(JavaTimeModule())

    @Cacheable("Bookings")
    fun getList(): List<Booking> {
        var token: String? = null
        val bookings = mutableListOf<Booking>()

        do {
            val request = calendarService.events().list(options.calendarId)
            if (token != null) {
                request.pageToken = token
            }

            val items = request.execute()
            token = items.nextPageToken

            val converted = items.items.mapNotNull { item ->
                try {
                    val data = item.extendedProperties?.shared?.get("Data") ?: return@mapNotNull null
                    val bytes = Base64.getDecoder().decode(data)

                    val booking: Booking = try {
                        objectMapper.readValue(String(bytes), Booking::class.java)
                    } catch (e: Exception) {
                        // Fallback: Extract from Google Event attributes since MemoryPack is binary
                        val b = Booking()
                        b.id = try { UUID.fromString(item.id) } catch (ex: Exception) { UUID.randomUUID() }
                        b.facilityName = item.location

                        // Parse description
                        // Point of contact: {pocName} / {pocPhone}
                        // Booked by: {unit} / {name}
                        // Number: {userPhone}
                        // Description: \n{description}
                        val desc = item.description ?: ""
                        val pocRegex = Regex("Point of contact: (.*) / (.*)")
                        val pocMatch = pocRegex.find(desc)
                        b.pocName = pocMatch?.groupValues?.get(1)?.trim()
                        b.pocPhone = pocMatch?.groupValues?.get(2)?.trim()

                        val numberRegex = Regex("Number: (.*)")
                        val numberMatch = numberRegex.find(desc)
                        b.userPhone = numberMatch?.groupValues?.get(1)?.trim()

                        val descriptionRegex = Regex("Description: \\s*(.*)", RegexOption.DOT_MATCHES_ALL)
                        val descriptionMatch = descriptionRegex.find(desc)
                        b.description = descriptionMatch?.groupValues?.get(1)?.trim()

                        // Summary: {unit} {conduct}
                        val summaryParts = item.summary?.split(" ", limit = 2)
                        b.conduct = if (summaryParts != null && summaryParts.size > 1) summaryParts[1] else item.summary

                        b
                    }

                    val eventStartDateTime = item.start?.dateTime?.value?.let {
                        OffsetDateTime.parse(item.start.dateTime.toStringRfc3339())
                    }
                    val eventEndDateTime = item.end?.dateTime?.value?.let {
                        OffsetDateTime.parse(item.end.dateTime.toStringRfc3339())
                    }

                    if (eventStartDateTime == null || eventEndDateTime == null) {
                        return@mapNotNull null
                    }

                    booking.startDateTime = eventStartDateTime
                    booking.endDateTime = eventEndDateTime
                    booking
                } catch (e: Exception) {
                    null
                }
            }

            bookings.addAll(converted)
        } while (token != null)

        return bookings
    }

    fun find(predicate: (Booking) -> Boolean): Booking? {
        return self.getList().find(predicate)
    }

    fun get(predicate: (Booking) -> Boolean): Booking {
        return self.getList().first(predicate)
    }

    @CacheEvict(value = ["Bookings"], allEntries = true)
    fun insert(entity: Booking): Booking {
        val user = userRepository.get { it.phone == entity.userPhone }

        val data = Base64.getEncoder().encodeToString(objectMapper.writeValueAsBytes(entity))

        if (data.length > 1000) {
            throw Exception("Event information is too long.")
        }

        val event = Event().apply {
            id = entity.id.toString().replace("-", "")
            summary = "${user.unit} ${entity.conduct}"
            start = EventDateTime().setDateTime(DateTime(entity.startDateTime.toString()))
            end = EventDateTime().setDateTime(DateTime(entity.endDateTime.toString()))
            location = entity.facilityName
            description = """
                Point of contact: ${entity.pocName} / ${entity.pocPhone}

                Booked by: ${user.unit} / ${user.name}
                Number: ${user.phone}

                Description:
                ${entity.description}
            """.trimIndent()
            extendedProperties = Event.ExtendedProperties().setShared(mapOf("Data" to data))
        }

        calendarService.events().insert(options.calendarId, event).execute()
        calendarService.events().insert(options.carbonCopyCalendarId, event).execute()

        return entity
    }

    @CacheEvict(value = ["Bookings"], allEntries = true)
    fun update(entity: Booking): Booking {
        val bookings = self.getList()
        val booking = bookings.single { it.id == entity.id }

        booking.startDateTime = entity.startDateTime
        booking.endDateTime = entity.endDateTime
        booking.conduct = entity.conduct
        booking.description = entity.description
        booking.facilityName = entity.facilityName
        booking.pocName = entity.pocName
        booking.pocPhone = entity.pocPhone
        booking.userPhone = entity.userPhone

        val data = Base64.getEncoder().encodeToString(objectMapper.writeValueAsBytes(booking))

        if (data.length > 1000) {
            throw Exception("Event information is too long.")
        }

        val user = userRepository.get { it.phone == booking.userPhone }

        val event = Event().apply {
            id = booking.id.toString().replace("-", "")
            summary = "${user.unit} ${booking.conduct}"
            start = EventDateTime().setDateTime(DateTime(booking.startDateTime.toString()))
            end = EventDateTime().setDateTime(DateTime(booking.endDateTime.toString()))
            location = booking.facilityName
            description = """
                Point of contact: ${booking.pocName} / ${booking.pocPhone}

                Booked by: ${user.unit} / ${user.name}
                Number: ${user.phone}

                Description:
                ${booking.description}
            """.trimIndent()
            extendedProperties = Event.ExtendedProperties().setShared(mapOf("Data" to data))
        }

        calendarService.events().update(options.calendarId, event.id, event).execute()
        calendarService.events().update(options.carbonCopyCalendarId, event.id, event).execute()

        return booking
    }

    @CacheEvict(value = ["Bookings"], allEntries = true)
    fun delete(predicate: (Booking) -> Boolean) {
        val booking = self.get(predicate)
        val eventId = booking.id.toString().replace("-", "")

        calendarService.events().delete(options.calendarId, eventId).execute()
        calendarService.events().delete(options.carbonCopyCalendarId, eventId).execute()
    }
}